hbase-server/src/main/java/org/apache/hadoop/hbase/master/janitor/CatalogJanitor.java

```diff
@@ -422,7 +426,16 @@ private static Pair<Boolean, Boolean> checkRegionReferences(MasterServices servi
     try {
       HRegionFileSystem regionFs = HRegionFileSystem
         .openRegionFromFileSystem(services.getConfiguration(), fs, tabledir, region, true);
-      boolean references = regionFs.hasReferences(tableDescriptor);
+      ColumnFamilyDescriptor[] families = tableDescriptor.getColumnFamilies();
+      boolean references = false;
+      for (ColumnFamilyDescriptor cfd : families) {
+        StoreFileTracker sft = StoreFileTrackerFactory.create(services.getConfiguration(),
+          tableDescriptor, ColumnFamilyDescriptorBuilder.of(cfd.getNameAsString()), regionFs);
+        references = references || sft.hasReferences();
+        if (references) {
+          break;
+        }
+      }
```

Comment:
Should also be moved to HRegionFileSystem.hasReferences(), like other methods did?

Additional question: Is it guaranteed that we don't have more than one SFT instance for the same store at any point in time? Because the file based SFT impl relies on the fact it's the single instance manipulating its meta files. If we have multiple instances of SFT for the same store, it could lead to inconsistencies, where one of the instances update the meta file, than the others would be looking at an outdated state of the meta files.

Reply:
> Should also be moved to HRegionFileSystem.hasReferences(), like other methods did?

When we introduce virtual links to Link/Ref files using SFT,  hasReferences() abstraction also should move to SFT layer and there is only one reference left of HRegionFileSystem.hasReferences() which will be eventually removed in [HBASE-28861](https://issues.apache.org/jira/browse/HBASE-28861) 

>  Is it guaranteed that we don't have more than one SFT instance for the same store at any point in time? Because the file based SFT impl relies on the fact it's the single instance manipulating its meta files

This should only be true for WRITE mode of SFT instance right? there could be multiple instances of READ only mode SFT for the same store. I have created HBASE-28863  for introducing cache primarily for READ only mode

>If we have multiple instances of SFT for the same store, it could lead to inconsistencies, where one of the instances update the meta file, than the others would be looking at an outdated state of the meta files.

Is this part not already handled based on timestamp? Or Flush and Compaction both are using SFT instance from StoreEngine due to that same reason?
Also looks like at couple of places I am using non READ only mode and I was only needing READ mode there, let me quickly reiterate on that, but I guess it should be good to have that check at StoreFileTrackerFactory layer as well. thanks for pointing that out @wchevreuil

Reply:
>This should only be true for WRITE mode of SFT instance right? there could be multiple instances of READ only mode SFT for the same store.

How would the "READ_ONLY" SFTs know that the "WRITER" had rotated the file? Specially if its on a different process, like here? Other master based processes such as split take it for granted since the region would be closed, but I don't think it's the case here.

Reply:
I was looking at the `FileBasedStoreFileTracker` code. Whenever compaction or flush happens, we rotate the meta file as we call `FileBasedStoreFileTracker.update`, and then we also update the map of store files within this `FileBasedStoreFileTracker` instance. At this point, any other existing `FileBasedStoreFileTracker` instance should reload the store files, otherwise the store files map on those instances would be outdated. I don't know if we are doing that for read replicas somehow, otherwise, we have a problem for SFT and read replicas.

Reply:
>At this point, any other existing FileBasedStoreFileTracker instance for this same store should reload the store files, otherwise the store files map on those instances would be outdated. 

load of SFT every time does a [backedfile.load()](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/storefiletracker/FileBasedStoreFileTracker.java#L71) again right, at that time it should know the latest manifest and load that one, it does not return the storefiles it cached know